### ISSUE TITLE
Add support for filtering Java 8 Streams with @PostFilter

### ISSUE THREAD

**Issue by shollander** [spring-projects/spring-security#3743]
It would be very useful to be able to `@PostFilter` methods that return a `java.util.stream.Stream` object. It should be relatively simple to add something like this to [DefaultMethodSecurityExpressionHandler](../blob/master/core/src/main/java/org/springframework/security/access/expression/method/DefaultMethodSecurityExpressionHandler.java)`.filter()`

```
if (filterTarget instanceof Stream) {
    Stream<?> filtered = (Stream<?>) filterTarget;
    filtered = filtered.filter(filterObject -> {
        rootObject.setFilterObject(filterObject);
        return ExpressionUtils.evaluateAsBoolean(filterExpression, ctx);
    });
    return filtered;
}
```

**Comment by nightswimmings** [comment]
I am not sure this is a good idea... Since filter() is an intermediate operation you are returning a new stream. How do we close the original one?

**Comment by kagof** [comment]
I'd like to claim this, if nobody else has yet!

**Comment by rwinch** [comment]
Thanks @kagof It is all yours!

**Comment by edeandrea** [comment]
I'm working on another PR and have come across this change. Isn't this line from DefaultMethodSecurityExpressionHandler a race condition?

return original.filter(filterObject -> {			rootObject.setFilterObject(filterObject);
				return ExpressionUtils.evaluateAsBoolean(filterExpression, ctx);
			})

The EvaluationContext and therefore the root object is a shared resource and setting the filter object on the root object could be happening in parallel so therefore the expression being evaluated could be using the wrong filter object.

I don't see how this code guarantees that the filter object being used in the expression evaluation is the correct one if the stream is a parallel stream.

**Comment by jzheaux** [comment]
@edeandrea I agree that if the `Stream` instance is parallelized, then calls to the `filter`'s lambda may happen in parallel, all operating on the same `MethodSecurityExpressionOperations` instance.

Creating a new context for each object in the stream could be expensive, and so I'm inclined to either error when the stream is parallelized, e.g. `Parallel Streams not supported`, or have the user explicitly switch-on support, e.g. `setAllowParallelStreamFiltering`.

@rwinch do you have some thoughts on filtering parallel streams?

**Comment by edeandrea** [comment]
Thanks for the input @jzheaux. My fear though is what we are going to run into the same issue when supporting reactive types with filters. Can we really say we don't support pre/post filtering on reactive types?

**Comment by rwinch** [comment]
I think at this point we need to create a new context for each object. It may be costly, but it is something we can optimize later if we have issues with it.
